File: learn/ruby/usingjruby/java/OverloadDemo.java
import java.util.List;
public class OverloadDemo {
    public static String whatTypeIs(long value) {
        return "long";
    }
    public static String whatTypeIs(String value) {
        return "string";
    }
    public static String whatTypeIs(Object value) {
        return "object";
    }

    public int neededFor(int i) {
        return 32;
    }
    public int neededFor(long l) {
        return 64;
    }
}

File: learn/ruby/usingjruby/lib/javafromruby/overload_demo.rb
Demo = Java::OverloadDemo
puts Demo.what_type_is(42)
puts Demo.what_type_is('Fun!')
puts Demo.what_type_is(Hash.new)


od = Demo.new
# jruby 你默认转换为long
puts od.needed_for(10_10000) # long

# 使用java_send，强制jruby使用int
puts (od.java_send :neededFor, [Java::int], 10_10000)

# 打开类 定义java方法别名
class Java::OverloadDemo
  java_alias :needed_for_int, :neededFor, [Java::int]
end
puts od.needed_for_int(1_000_000) # -> int


# 使用java_method 引用重载的方法
bits = Demo.new
bits_needed_for = bits.java_method :neededFor, [Java::int]
puts (bits_needed_for.call 1_000_000) # -> int